contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -103,20 +121,70 @@ fn ensure_its_source_address(
     Ok(())
 }
 
+fn verify_coin(
+    deps: &DepsMut,
+    coin: &Option<Coin>,
+    message: &Message,
+    source_chain: &ChainNameRaw,
+) -> Result<(), Error> {
+    let expected_coin = gateway_token_transfer(deps, source_chain, message)?;
+    ensure!(
+        &expected_coin == coin,
+        Error::IncorrectAttachedCoin {
+            attached: coin.clone(),
+            expected: expected_coin
+        }
+    );
+    Ok(())
+}
+
+fn gateway_token_transfer(
+    deps: &DepsMut,
+    source_chain: &ChainNameRaw,
```

Comment:
```suggestion
    chain: &ChainNameRaw,
```
since it could be source/destination depending on context. Could also call it `core_chain` but we haven't verified that if that's the case yet